components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2470,6 +2488,46 @@ where
         self.fsm.hibernate_state.count_vote(from.get_id());
     }
 
+    fn on_collect_peer_apply_index_process_request(&mut self, from: &metapb::Peer) {
+        if self.fsm.peer.is_leader() {
+            return;
+        }
+        let mut msg = ExtraMessage::default();
+        msg.set_type(ExtraMessageType::MsgCollectPeerApplyIndexProgressResponse);
+        msg.applied_index = self.fsm.peer.get_store().applied_index();
+        self.fsm
+            .peer
+            .send_extra_message(msg, &mut self.ctx.trans, from);
+    }
+
+    fn on_collect_peer_apply_index_process_response(
+        &mut self,
+        from: &metapb::Peer,
+        applying_idx: u64,
+    ) {
+        if !self.fsm.peer.is_leader() {
+            return;
+        }
+        if self
+            .fsm
+            .peer
+            .region()
+            .get_peers()
+            .iter()
+            .all(|p| p.get_id() != from.get_id())
+        {
+            self.fsm.peer.peer_applied_indices.remove(&from.get_id());
+            return;
+        }
+        let prev_applying_idx = self
+            .fsm
+            .peer
+            .peer_applied_indices
+            .entry(from.get_id())
+            .or_insert(applying_idx);
+        *prev_applying_idx = applying_idx;
+    }
+
     fn on_extra_message(&mut self, mut msg: RaftMessage) {
```

Comment:
If there is an extra message, it would return directly after handling the extra message which leaves the raft message unresolved. Should change the spot of `has_extra_message` as well

Reply:
Oh indeed. How about we modify the `on_extra_message` function signature to add a return value? 
- Return `false`, indicating only extra msg and no raft msg to process, return directly as before
- Return `true`, indicating there is also raft msg to be processed and cannot be returned directly

Reply:
Oh indeed. How about we modify the signature of `on_extra_message` to add a return value?

- return `false`, it means that there is only extra msg without raft msg or snapshot, etc., it would return directly as before.
- return `true`, indicating that there are still raft msg or snapshot, etc. to process. 

As most `ExtraMessageType` does not need to execute the code behind `on_extra_message`.

Reply:
we simply do not return after `on_extra_message`, and check `has_message` below

Reply:
Sure, fixed thanks!